File: zig/src/game_of_life.zig
const std = @import("std");
const Benchmark = @import("benchmark.zig").Benchmark;
const Helper = @import("helper.zig").Helper;

pub const GameOfLife = struct {
    allocator: std.mem.Allocator,
    helper: *Helper,
    width: usize,
    height: usize,
    cells: [][]Cell,

    const Cell = struct {
        alive: bool,
        next_state: bool,
        neighbors: [8]*Cell,
        neighbor_count: usize,

        fn init() Cell {
            return Cell{
                .alive = false,
                .next_state = false,
                .neighbors = undefined,
                .neighbor_count = 0,
            };
        }

        fn addNeighbor(self: *Cell, neighbor: *Cell) void {
            self.neighbors[self.neighbor_count] = neighbor;
            self.neighbor_count += 1;
        }

        fn computeNextState(self: *Cell) void {
            var alive_neighbors: usize = 0;
            for (self.neighbors) |n| {
                if (n.alive) alive_neighbors += 1;
            }

            if (self.alive) {
                self.next_state = alive_neighbors == 2 or alive_neighbors == 3;
            } else {
                self.next_state = alive_neighbors == 3;
            }
        }

        fn update(self: *Cell) void {
            self.alive = self.next_state;
        }
    };

    const vtable = Benchmark.VTable{
        .run = runImpl,
        .checksum = checksumImpl,
        .prepare = prepareImpl,
        .deinit = deinitImpl,
    };

    pub fn init(allocator: std.mem.Allocator, helper: *Helper) !*GameOfLife {
        const w = helper.config_i64("GameOfLife", "w");
        const h = helper.config_i64("GameOfLife", "h");
        const width = @as(usize, @intCast(w));
        const height = @as(usize, @intCast(h));

        var cells = try allocator.alloc([]Cell, height);
        for (0..height) |y| {
            cells[y] = try allocator.alloc(Cell, width);
            for (0..width) |x| {
                cells[y][x] = Cell.init();
            }
        }

        const self = try allocator.create(GameOfLife);
        self.* = GameOfLife{
            .allocator = allocator,
            .helper = helper,
            .width = width,
            .height = height,
            .cells = cells,
        };

        try self.linkNeighbors();
        return self;
    }

    fn linkNeighbors(self: *GameOfLife) !void {
        for (0..self.height) |y| {
            for (0..self.width) |x| {
                const cell = &self.cells[y][x];

                var dy: i32 = -1;
                while (dy <= 1) : (dy += 1) {
                    var dx: i32 = -1;
                    while (dx <= 1) : (dx += 1) {
                        if (dx == 0 and dy == 0) continue;

                        const ny = @mod(@as(i32, @intCast(y)) + dy + @as(i32, @intCast(self.height)), @as(i32, @intCast(self.height)));
                        const nx = @mod(@as(i32, @intCast(x)) + dx + @as(i32, @intCast(self.width)), @as(i32, @intCast(self.width)));

                        const neighbor = &self.cells[@as(usize, @intCast(ny))][@as(usize, @intCast(nx))];
                        cell.addNeighbor(neighbor);
                    }
                }
            }
        }
    }

    fn nextGeneration(self: *GameOfLife) void {
        for (self.cells) |row| {
            for (row) |*cell| {
                cell.computeNextState();
            }
        }

        for (self.cells) |row| {
            for (row) |*cell| {
                cell.update();
            }
        }
    }

    fn countAlive(self: *const GameOfLife) u32 {
        var count: u32 = 0;
        for (self.cells) |row| {
            for (row) |cell| {
                if (cell.alive) count += 1;
            }
        }
        return count;
    }

    fn computeHash(self: *const GameOfLife) u32 {
        const FNV_OFFSET_BASIS: u32 = 2166136261;
        const FNV_PRIME: u32 = 16777619;

        var hash: u32 = FNV_OFFSET_BASIS;
        for (self.cells) |row| {
            for (row) |cell| {
                const alive: u32 = if (cell.alive) 1 else 0;
                hash ^= alive;
                hash *%= FNV_PRIME;
            }
        }
        return hash;
    }

    pub fn deinit(self: *GameOfLife) void {
        for (0..self.height) |y| {
            self.allocator.free(self.cells[y]);
        }
        self.allocator.free(self.cells);
        self.allocator.destroy(self);
    }

    pub fn asBenchmark(self: *GameOfLife) Benchmark {
        return Benchmark.init(self, &vtable, self.helper, "GameOfLife");
    }

    fn prepareImpl(ptr: *anyopaque) void {
        const self: *GameOfLife = @ptrCast(@alignCast(ptr));
        for (self.cells) |row| {
            for (row) |*cell| {
                if (self.helper.nextFloat(1.0) < 0.1) {
                    cell.alive = true;
                }
            }
        }
    }

    fn runImpl(ptr: *anyopaque, _: i64) void {
        const self: *GameOfLife = @ptrCast(@alignCast(ptr));
        self.nextGeneration();
    }

    fn checksumImpl(ptr: *anyopaque) u32 {
        const self: *GameOfLife = @ptrCast(@alignCast(ptr));
        return self.computeHash() + self.countAlive();
    }

    fn deinitImpl(ptr: *anyopaque) void {
        const self: *GameOfLife = @ptrCast(@alignCast(ptr));
        self.deinit();
    }
};

File: csharp/GameOfLife.cs
public class GameOfLife : Benchmark
{
    private class Cell
    {
        public bool Alive { get; set; }
        public bool NextState { get; set; }
        private Cell[] _neighbors = new Cell[8];
        private int _neighborCount;

        public void AddNeighbor(Cell cell) => _neighbors[_neighborCount++] = cell;

        public void ComputeNextState()
        {
            int aliveNeighbors = 0;
            foreach (var neighbor in _neighbors)
            {
                if (neighbor.Alive) aliveNeighbors++;
            }

            NextState = Alive
                ? aliveNeighbors == 2 || aliveNeighbors == 3
                : aliveNeighbors == 3;
        }

        public void Update() => Alive = NextState;
    }

    private class Grid
    {
        private readonly int _width;
        private readonly int _height;
        private readonly Cell[][] _cells;

        public Grid(int width, int height)
        {
            _width = width;
            _height = height;
            _cells = new Cell[height][];

            for (int y = 0; y < height; y++)
            {
                _cells[y] = new Cell[width];
                for (int x = 0; x < width; x++)
                    _cells[y][x] = new Cell();
            }

            LinkNeighbors();
        }

        private void LinkNeighbors()
        {
            for (int y = 0; y < _height; y++)
            {
                for (int x = 0; x < _width; x++)
                {
                    var cell = _cells[y][x];

                    for (int dy = -1; dy <= 1; dy++)
                    {
                        for (int dx = -1; dx <= 1; dx++)
                        {
                            if (dx == 0 && dy == 0) continue;

                            int ny = (y + dy + _height) % _height;
                            int nx = (x + dx + _width) % _width;

                            cell.AddNeighbor(_cells[ny][nx]);
                        }
                    }
                }
            }
        }

        public void NextGeneration()
        {

            foreach (var row in _cells)
                foreach (var cell in row)
                    cell.ComputeNextState();

            foreach (var row in _cells)
                foreach (var cell in row)
                    cell.Update();
        }

        public int CountAlive()
        {
            int count = 0;
            foreach (var row in _cells)
                foreach (var cell in row)
                    if (cell.Alive) count++;
            return count;
        }

        public uint ComputeHash()
        {
            const uint FNV_OFFSET_BASIS = 2166136261u;
            const uint FNV_PRIME = 16777619u;

            uint hash = FNV_OFFSET_BASIS;
            foreach (var row in _cells)
                foreach (var cell in row)
                {
                    uint alive = cell.Alive ? 1u : 0u;
                    hash = (hash ^ alive) * FNV_PRIME;
                }
            return hash;
        }

        public Cell[][] GetCells() => _cells;
    }

    private readonly int _width;
    private readonly int _height;
    private Grid _grid;

    public GameOfLife()
    {
        _width = (int)ConfigVal("w");
        _height = (int)ConfigVal("h");
        _grid = new Grid(_width, _height);
    }

    public override void Prepare()
    {
        foreach (var row in _grid.GetCells())
            foreach (var cell in row)
                if (Helper.NextFloat() < 0.1f)
                    cell.Alive = true;
    }

    public override void Run(long iterationId) => _grid.NextGeneration();

    public override uint Checksum
    {
        get
        {
            int alive = _grid.CountAlive();
            return _grid.ComputeHash() + (uint)alive;
        }
    }
}